### PR TITLE
Jacobian coordinates

### LINKED ISSUE #6: Introduce other point coordinate systems
We currently have only Affine and Projective coordinates logic for the curve arithmetic.
Being able to support other point coordinate systems may bring some performance benefits, for instance in the [schnorr-sig](https://github.com/Toposware/schnorr-sig) crate.

For instance Jacobian coordinates are known for yielding relatively fast point doubling methods (about 2/3 of the running time of a point doubling in Projective coordinates) for a slightly slower addition algorithm. However, the use of Straus-Shamir's trick in signature verification reduces the impact of such change (as we have now N doublings and 2N additions).

Jacobian coordinates variants (like Chudnovsky) could also be investigated, though at the cost of a heavier point representation (240B instead of 144B for Jacobian/Projective).

### PR DESCRIPTION
**Pull request by Nashtare**
This PR introduces Jacobian (normal and Modified ones) coordinates, and closes #6.

Compared to projective scalar multiplication, modified jacobian scalar multiplication yields on my machine a speed-up of:
- 18.6% for constant-time variable-based scalar multiplication
- 22.2% for variable-time variable-based scalar multiplication (for identical scalar with ~ half bits set to 1)
- 11.5% for constant-time variable-based double scalar multiplication
- 9.7% for variable-time variable-based double scalar multiplication (for 2 identical scalars with ~ half bits set to 1)
- 17.4% for variable-time double scalar multiplication with basepoint

Regular scalar multiplication with a hardcoded basepoint via the Pippenger algorithm is pretty slow, as there is only 4 doublings. The introduction of a specially hardcoded BASEPOINT_LOOKUP allows for faster scalar multiplication with the Cheetah basepoint in jacobian coordinates, compared to the variable-based one, but is still much slower than the one in Projective coordinates (about ~3x slower).

### CODE REVIEW

File: src/curve/affine.rs [hunk 1]
```diff
@@ -0,0 +1,686 @@
...
+
+        point.is_torsion_free()
+    }
+}
+
+impl<'a, 'b> Mul<&'b Scalar> for &'a AffinePoint {
+    type Output = JacobianPoint;
+
+    fn mul(self, other: &'b Scalar) -> Self::Output {
+        JacobianPoint::from(self).multiply(&other.to_bytes())
```
Comment (baumbata) [inline src/curve/affine.rs:424]: Is there a reason to prefer this form over `JacobianPoint::multiply(&self.into(), &other.to_bytes())`?

Comment (baumbata) [reply src/curve/affine.rs:424]: No, just noticed it being different from e.g. `JacobianPoint::multiply_double_with_basepoint_vartime(&self.into(), by_self, by_basepoint)` on line 400 above and wondered if there was an advantage I was missing.

File: src/curve/encoding.rs [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+        );
+
+        // Is this point at infinity? If so, set the most significant bit of the last byte.
+        bytes[96] |= u8::conditional_select(&0u8, &(1u8 << 7), point.infinity);
+
+        Self(bytes)
+    }
+
+    /// Attempts to convert an `UncompressedPoint` to an `AffinePoint`
+    /// The resulting point is ensured to be on the curve, but is
```
Comment (baumbata) [inline src/curve/encoding.rs:176]: Is this point actually ensured to be on the curve? It seems we just recover x and y as given.

Comment (Nashtare) [reply src/curve/encoding.rs:176]: This is an incorrect leftover from some previous PR. Indeed, for `UncompressedPoint`, as there is no `y` recovery, the additional curve check must be done outside of this method. This is done on the `AffinePoint` method `from_uncompressed`, on which other points coordinate methods are based. Conversion from a `CompressedPoint` already includes this check and should hence only check that we are on the correct subgroup. I'll correct it, thanks!

File: src/curve/jacobian.rs [hunk 3]
```diff
@@ -0,0 +1,1859 @@
...
+            output = output.double();
+        }
+
+        output
+    }
+
+    /// Computes `n` iterated doubling of this point.
+    /// **This is dangerous to call unless you know that the point to be doubled
+    /// is not the identity point, otherwise, API invariants may be broken.**
+    /// Please consider using `double()` instead.
```
Comment (baumbata) [inline src/curve/jacobian.rs:353]: Minor question: should it be "Please consider using `double_multi()` instead." here?

**Pull request merged**